packages/angular/cli/src/utilities/memoize.ts

```diff
@@ -0,0 +1,40 @@
+/**
+ * @license
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+/** A decorator that memoizes methods and getters. */
+export function memoize(): MethodDecorator {
+  return <T>(
+    _target: Object,
+    _propertyKey: string | symbol,
+    descriptor: TypedPropertyDescriptor<T>,
+  ) => {
+    const descriptorPropertyName = descriptor.get ? 'get' : 'value';
+    const originalMethod: unknown = descriptor[descriptorPropertyName];
+
+    if (typeof originalMethod !== 'function') {
+      throw new Error('memoize decorator can only be used on methods or get accessors.');
+    }
+
+    const cache = new Map<string, unknown>();
+
+    return {
+      ...descriptor,
+      [descriptorPropertyName]: function (this: unknown, ...args: unknown[]) {
+        const key = JSON.stringify(args);
```

Comment:
This requires that arguments are serializable to JSON. A quick look at the [algorithm's explanation (unofficial)](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify#description) shows that non-serializable arguments (such as a function) return `'null'`, which could lead to surprising bugs. Ideally we wouldn't have a serializability requirement and I think that's *possible* with some kind of recursive `Map<first param, Map<second param, Map<..., result>>>`.

That would be pretty complicated, so I think we should at least document the serializability requirement and possibly check the arguments to make sure they are serializable. For example, we could `JSON.stringify()` each arg and make sure the result is only `'null'` if the actual value is `null`. There might be other cases, and getters might still cause confusion, but that would probably be "good enough" safety for what we're doing here.

Reply:
Updated. PTAL